fix: handle error in numberToPos and formatError

milahu opened the pull request:
### old behavior

1. formatError is trying to handle error1
2. formatError calls numberToPos
3. numberToPos throws error2
4. only error2 is printed
5. error1 is not printed

for example

```
Error: offset is longer than source length!
    at numberToPos (node_modules/vite/dist/node/chunks/dep-972722fa.js:4247:15)
    at formatError (node_modules/vite/dist/node/chunks/dep-972722fa.js:51169:24)
    at TransformContext.error (node_modules/vite/dist/node/chunks/dep-972722fa.js:51149:19)
    at Object.transform (node_modules/vite/dist/node/chunks/dep-972722fa.js:51357:25)
    at async transformRequest (node_modules/vite/dist/node/chunks/dep-972722fa.js:67098:29)
    at async instantiateModule (node_modules/vite/dist/node/chunks/dep-972722fa.js:73732:10)
```

### new behavior

```
formatError: error in numberToPos (Error: offset is longer than source length! offset 126 > length 48) while handling the error:
Expected valid tag name
4: </script>
5: 
6: <p>Lorem ipsum < dolor sit</p>
                   ^
ParseError: Expected valid tag name
    at error (node_modules/svelte/compiler.mjs:17698:19)
    at Parser$1.error (node_modules/svelte/compiler.mjs:17774:9)
    [...]
```

---

### What is the purpose of this pull request? <!-- (put an "X" next to an item) -->

- [x] Bug fix
- [ ] New Feature
- [ ] Documentation update
- [ ] Other

### Before submitting the PR, please make sure you do the following

- [x] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md).
- [x] Read the [Pull Request Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md#pull-request-guidelines) and follow the [Commit Convention](https://github.com/vitejs/vite/blob/main/.github/commit-convention.md).
- [x] Check that there isn't already a PR that solves the problem the same way to avoid creating a duplicate.
- [x] Provide a description in this PR that addresses **what** the PR is solving, or reference the issue that it solves (e.g. `fixes #123`).
- [ ] Ideally, include relevant tests that fail without this PR but pass with it.

Shinigami92 commented on packages/vite/src/node/server/pluginContainer.ts:
```diff
@@ -285,10 +285,18 @@ export async function createPluginContainer(
+          errLocation = numberToPos(ctx._activeCode, pos);
         }
+        catch (err2) {
+          console.log(`formatError: error in numberToPos (${err2}) while handling the error:`);
```
I think you need to use the logger

milahu committed:
fix: handle errors in formatError and numberToPos (#4782)

milahu committed:
fix: remove dependency acorn-numeric-separator (#4782)

deprecated acorn-numeric-separator@0.3.6: acorn>=7.4 supports numeric separators

Shinigami92 commented on package.json:
```diff
@@ -54,7 +54,7 @@
     "ts-jest": "^27.0.4",
     "ts-node": "^10.1.0",
-    "typescript": "^4.3.5",
+    "typescript": "4.3.5",
```
You may need to use `yarn install --frozen-lockfile`
Please do not alter `package.json` within this PR

milahu replied:
sorry for the mess, should have read contributing.md earlier

Shinigami92 replied:
Please revert this change

Shinigami92 commented on packages/vite/package.json:
```diff
@@ -79,7 +79,6 @@
     "@vue/compiler-dom": "^3.1.5",
     "acorn": "^8.4.1",
     "acorn-class-fields": "^1.0.0",
-    "acorn-numeric-separator": "^0.3.6",
```
**question:** Why is this not required anymore? What does it has to do with this PR?
Or is it really just used at these lines and can therefore be removed?

milahu replied:
see commit message:

> fix: remove dependency acorn-numeric-separator (#4782)
>
> deprecated acorn-numeric-separator@0.3.6: acorn>=7.4 supports numeric separators

milahu commented on packages/vite/src/node/server/pluginContainer.ts:
```diff
@@ -285,10 +283,22 @@ export async function createPluginContainer(
-        err.loc = err.loc || {
-          file: err.id,
-          ...numberToPos(ctx._activeCode, pos)
+        let errLocation;
+        try {
+          errLocation = numberToPos(ctx._activeCode, pos);
         }
+        catch (err2) {
+          logger.error(
+            chalk.red(`Error in error handler:\n${err2.stack || err2.message}`),
+            { error: err2 }
+          )
+          // TODO verify that err is visible
+          throw err;
+        }
```
> TODO verify that err is visible

milahu replied:
how to add a unit test for this?
vite.createServer -> createPluginContainer -> formatError

broken code is passed to the plugin -> plugin throws error1
plugin returns a broken sourcemap -> numberToPos throws error2 "offset is longer than source length"

Shinigami92 replied:
@patak-js Do we even test logged errors?
I'm not sure if we need to test this :thinking:

Shinigami92 commented on yarn.lock:
```diff
@@ -7722,7 +7717,7 @@ typedarray-to-buffer@^3.1.5:
   uid ""
 
-typescript@^4.3.4, typescript@^4.3.5, typescript@~4.3.5:
+typescript@4.3.5, typescript@^4.3.4, typescript@~4.3.5:
```
Please revert this change

milahu replied:
this is temporary until https://github.com/rollup/plugins/issues/983 is fixed

Shinigami92 commented:
Converted to draft until `TODO verify that err is visible` is resolved

milahu commented:
> TODO verify that err is visible

done manually, this is non-trivial to test

<details>

<pre>
$ pnpm run dev 

> @ dev svelte/mdsvex/sveltekit-bugs
> svelte-kit dev

<span color="cyan">Pre-bundling dependencies:</span>
<span color="orange">  svelte/store
  svelte
  svelte/animate
  svelte/easing
  svelte/internal
  (...and 2 more)</span>
(this will be run only when your dependencies or config have changed)

  <span color="cyan"><b>SvelteKit v1.0.0-next.159</b></span>

  local:   http://localhost:3000
  network: not exposed

  Use --host to expose server to other devices on this network


<span color="red">Error in error handler:
Error: offset is longer than source length! offset 126 > length 48
    at numberToPos (vite/dist/node/chunks/dep-7017d082.js:4256:15)
    at formatError (vite/dist/node/chunks/dep-7017d082.js:50004:35)
    at TransformContext.error (vite/dist/node/chunks/dep-7017d082.js:49984:19)
    at Object.transform (vite/dist/node/chunks/dep-7017d082.js:50200:25)
    at async transformRequest (vite/dist/node/chunks/dep-7017d082.js:64964:29)
    at async instantiateModule (vite/dist/node/chunks/dep-7017d082.js:74748:10)
<b>Expected valid tag name</b></span>
4: &lt;/script>
5: 
6: &lt;p>Lorem ipsum < dolor sit&lt;/p>
                   ^
ParseError: Expected valid tag name
    at error (svelte/compiler.mjs:17697:19)
    at Parser$1.error (svelte/compiler.mjs:17773:9)
    at read_tag_name (svelte/compiler.mjs:16918:16)
    at tag (svelte/compiler.mjs:16768:18)
    at new Parser$1 (svelte/compiler.mjs:17732:21)
    at parse$3 (svelte/compiler.mjs:17864:20)
    at compile (svelte/compiler.mjs:31303:17)
    at compileSvelte (vite-plugin-svelte/dist/index.js:301:20)
    at async TransformContext.transform (vite-plugin-svelte/dist/index.js:1325:27)
    at async Object.transform (vite/dist/node/chunks/dep-7017d082.js:50197:30)
</pre>

let me just add a newline after the first error ...

</details>

milahu commented:
note: currently, the browser runtime shows only the second error
both errors are visible only in the terminal
